Sources/Algorithms/LazySplitCollection.swift

```diff
@@ -0,0 +1,367 @@
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+/// A collection that lazily splits a base collection into subsequences
+/// separated by elements that satisfy the given `whereSeparator` predicate.
+///
+/// - Note: This type is the result of
+///
+///     x.split(maxSplits:omittingEmptySubsequences:whereSeparator)
+///     x.split(separator:maxSplits:omittingEmptySubsequences)
+///
+///   where `x` conforms to `LazyCollectionProtocol`.
+public struct LazySplitCollection<Base: Collection> {
+  internal let base: Base
+  internal let isSeparator: (Base.Element) -> Bool
+  internal let maxSplits: Int
+  internal let omittingEmptySubsequences: Bool
+  internal var _startIndex: Index
+
+  internal init(
+    base: Base,
+    isSeparator: @escaping (Base.Element) -> Bool,
+    maxSplits: Int,
+    omittingEmptySubsequences: Bool
+  ) {
+    self.base = base
+    self.isSeparator = isSeparator
+    self.maxSplits = maxSplits
+    self.omittingEmptySubsequences = omittingEmptySubsequences
+    self._startIndex = Index(baseRange: base.startIndex..<base.startIndex)
+
+    if !base.isEmpty {
+      // Precompute the start index.
+      _startIndex = indexForSubsequence(atOrAfter: base.startIndex)
+    }
+  }
+}
+
+extension LazySplitCollection: LazyCollectionProtocol {
+  /// Position of a subsequence in a split collection.
+  public struct Index: Comparable {
+    /// The range corresponding to the subsequence at this position.
+    internal let baseRange: Range<Base.Index>
+    /// The number of subsequences up to and including this position in the
+    /// collection.
+    internal let sequenceLength: Int
+    internal let separatorCount: Int
+
+    internal init(
+      baseRange: Range<Base.Index>,
+      sequenceLength: Int = 0,
+      separatorCount: Int = 0
+    ) {
+      self.baseRange = baseRange
+      self.sequenceLength = sequenceLength
+      self.separatorCount = separatorCount
+    }
+
+    public static func == (lhs: Index, rhs: Index) -> Bool {
+      // Since each index represents the range of a disparate subsequence, no
+      // two unique indices will have the same lower bound.
+      lhs.baseRange.lowerBound == rhs.baseRange.lowerBound
+    }
+
+    public static func < (lhs: Index, rhs: Index) -> Bool {
+      // Only use the lower bound to test for ordering, as above.
+      lhs.baseRange.lowerBound < rhs.baseRange.lowerBound
+    }
+  }
+
+  /// Returns the index of the subsequence starting at or after the given base collection index.
+  internal func indexForSubsequence(
+    atOrAfter lowerBound: Base.Index,
+    sequenceLength: Int = 0,
+    separatorCount: Int = 0
```

Comment:
These defaults seem a bit risky. Can we be explicit at the call sites instead about what these counters are for the start and end index?

Reply:
Agreed, yes.

`separatorCount` is another bit of this index implementation I don't feel great about. It's only correct for all indices when we're not omitting empty subsequences, or when there aren't multiple adjacent separators. Otherwise, we explicitly skip over multiple adjacent separators without counting them, for a more succinct implementation. We don't need an accurate separator count when we're omitting empty subsequences, but it still bugs me that the property is there and not correct.

On the other hand, these are internal properties, and the initializer is internal, so I can view them as implementation details and feel somewhat better. Do you have any concerns about this?

Reply:
I am thinking the index will become an enum though, as discussed above. Then the end case wouldn't have these properties at all.